modules/contact/src/MechanicalContactConstraint.C

```diff
@@ -126,8 +129,19 @@ MechanicalContactConstraint::MechanicalContactConstraint(const InputParameters &
     _mesh_dimension(_mesh.dimension()),
     _vars(3, libMesh::invalid_uint),
     _var_objects(3, nullptr),
+    _has_slave_gap_offset(isParamValid("slave_gap_offset")),
+    _slave_gap_offset_var(
+        _has_slave_gap_offset
+            ? &_fe_problem.getStandardVariable(_tid, getParam<VariableName>("slave_gap_offset"))
```

Comment:
Your assertion error is happening because of this line. You are retrieving the non-displaced copy of this variable (there are always displaced and non-displaced copies of variables when running with displaced mesh), and then later trying to retrieve the value of the variable on a ***displaced*** node. You get the failed semi-local assertion because the non-displaced variable copy is querying the non-displaced mesh for a displaced node, e.g. the non-displaced mesh has no idea about that displaced node.

Now this works in non-debug modes because the dof numbering and solution vectors are exactly the same between non-displaced and displaced configurations.

Still I would recommend the `getVar` API from the `Coupleable` class to retrieve the "correct" copy of the gap offset variables. This would also require that you add the gap offsets to your `InputParameters` using `params.addCoupledVar` as opposed to `params.addParam<VariableName>`. I think the "coupled" APIs are the better ones to use.